packages/twenty-front/src/modules/object-record/record-index/options/hooks/useExportTableData.ts

```diff
@@ -0,0 +1,157 @@
+import { useEffect, useState } from 'react';
+import { json2csv } from 'json-2-csv';
+import { useRecoilValue } from 'recoil';
+
+import { useFindManyRecords } from '@/object-record/hooks/useFindManyRecords';
+import { FieldMetadata } from '@/object-record/record-field/types/FieldMetadata';
+import { useRecordTableStates } from '@/object-record/record-table/hooks/internal/useRecordTableStates';
+import { ColumnDefinition } from '@/object-record/record-table/types/ColumnDefinition';
+
+import { useFindManyParams } from '../../hooks/useLoadRecordIndexTable';
+
+export const sleep = (ms: number) =>
+  new Promise((resolve) => setTimeout(resolve, ms));
+
+export const download = (blob: Blob, filename: string) => {
+  const url = URL.createObjectURL(blob);
+  const link = document.createElement('a');
+  link.href = url;
+  link.setAttribute('download', filename);
+  document.body.appendChild(link);
+  link.click();
+  link.parentNode?.removeChild(link);
+};
+
+type GenerateExportOptions = {
+  columns: ColumnDefinition<FieldMetadata>[];
+  rows: object[];
+};
+
+type GenerateExport = (data: GenerateExportOptions) => string;
+
+export const generateCsv: GenerateExport = ({
+  columns,
+  rows,
+}: GenerateExportOptions): string => {
+  const keys = columns.flatMap((col) => {
+    const column = {
+      field: col.metadata.fieldName,
+      title: col.label,
+    };
+
+    const objectColumnData = rows.find((r) => {
+      const val = (r as any)[column.field];
+      return val && typeof val === 'object' && !Array.isArray(val);
+    });
+
+    if (objectColumnData) {
+      const nestedFieldsWithoutTypename = Object.keys(
+        (objectColumnData as any)[column.field],
+      )
+        .filter((key) => key !== '__typename')
+        .map((k) => ({
+          field: `${column.field}.${k}`,
+          title: `${column.title}${k[0].toUpperCase() + k.slice(1)}`,
+        }));
+      return nestedFieldsWithoutTypename;
+    }
+    return [column];
+  });
+
+  return json2csv(rows, {
+    keys,
+    emptyFieldValue: '',
+  });
+};
+
+export const percentage = (part: number, whole: number): number => {
+  return Math.round((part / whole) * 100);
+};
+
+const downloader = (mimeType: string, generator: GenerateExport) => {
+  return (filename: string, data: GenerateExportOptions) => {
+    const blob = new Blob([generator(data)], { type: mimeType });
+    download(blob, filename);
+  };
+};
+
+export const csvDownloader = downloader('text/csv', generateCsv);
+
+type UseExportTableDataOptions = {
+  delayMs: number;
+  filename: string;
+  maximumRequests?: number;
+  objectNameSingular: string;
+  pageSize?: number;
+  recordIndexId: string;
+};
+
+export const useExportTableData = ({
```

Comment:
That works.

That's not critical right now, but we should handle the case where we have to stop at maximumRequests. 

I created an issue for that : https://github.com/twentyhq/twenty/issues/4161